File: src/main/java/com/theendercore/hydra/twitch/TwitchBot.java
package com.theendercore.hydra.twitch;

import com.github.philippheuer.credentialmanager.domain.OAuth2Credential;
import com.github.twitch4j.TwitchClientBuilder;
import com.github.twitch4j.chat.events.channel.ChannelMessageEvent;
import com.github.twitch4j.chat.events.channel.SubscriptionEvent;
import com.github.twitch4j.helix.domain.UserList;
import com.github.twitch4j.pubsub.events.FollowingEvent;
import com.github.twitch4j.pubsub.events.RewardRedeemedEvent;
import com.theendercore.hydra.config.ModConfig;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;

import java.util.List;
import java.util.Objects;

import static com.theendercore.hydra.HydraMod.*;
import static com.theendercore.hydra.util.Methods.chatMessage;

public class TwitchBot {

    public static int Enable() {
        ModConfig config = ModConfig.getConfig();
        credential = new OAuth2Credential("twitch", config.getOauthKey());
        if (Objects.equals(config.getUsername(), "") || Objects.equals(config.getOauthKey(), "")) {
            chatMessage(Text.translatable("command." + MODID + ".error.config").formatted(Formatting.RED));
            return 0;
        }
        if (twitchClient == null) {
            chatMessage(Text.translatable("command." + MODID + ".connecting", config.getUsername()).formatted(Formatting.DARK_GRAY));
            twitchClient = TwitchClientBuilder.builder().withEnableHelix(true).withEnablePubSub(true).withEnableChat(true).withChatAccount(credential).build();
            chatMessage(Text.translatable("command." + MODID + ".connected").formatted(Formatting.DARK_GRAY));
        } else {
            chatMessage(Text.translatable("command." + MODID + ".connected.already", config.getUsername()).formatted(Formatting.GRAY));
            return  0;
        }

        if (credential.getUserName() == null) {
            chatMessage(Text.translatable("command." + MODID + ".error.token.incorrect").formatted(Formatting.RED));
            Disable();
            return 0;
        }

        if (config.getExtras()) {
            UserList resultList = twitchClient.getHelix().getUsers(credential.getAccessToken(), null, List.of(config.getUsername())).execute();
            String channelID = resultList.getUsers().get(0).getId();
            LOGGER.info(channelID);
            twitchClient.getPubSub().listenForChannelPointsRedemptionEvents(credential, channelID);
            twitchClient.getPubSub().listenForFollowingEvents(credential, channelID);
            twitchClient.getPubSub().listenForSubscriptionEvents(credential, channelID);
            twitchClient.getEventManager().onEvent(RewardRedeemedEvent.class, EventListeners::rewardRedeemedListener);
            twitchClient.getEventManager().onEvent(FollowingEvent.class, EventListeners::followingEventListener);
            twitchClient.getEventManager().onEvent(SubscriptionEvent.class, EventListeners::subscriptionEventListener);
            chatMessage(Text.translatable("command." + MODID + ".extras.enable"));
        }

        twitchClient.getEventManager().onEvent(ChannelMessageEvent.class, EventListeners::channelMessageListener);
        return 1;
    }

    public static int Disable() {
        if (twitchClient == null) {
            chatMessage(Text.translatable("command." + MODID + ".disconnected.already").formatted(Formatting.DARK_GRAY));
            return 0;
        }
        twitchClient.getPubSub().close();
        twitchClient.getChat().close();
        twitchClient.close();
        twitchClient = null;
        chatMessage(Text.translatable("command." + MODID + ".disconnected").formatted(Formatting.GRAY));

        return 1;
    }
}

File: src/main/kotlin/com/theendercore/hydra/twitch/TwitchBot.kt
package com.theendercore.hydra.twitch

import com.github.philippheuer.credentialmanager.domain.OAuth2Credential
import com.github.twitch4j.TwitchClient
import com.github.twitch4j.TwitchClientBuilder
import com.github.twitch4j.chat.events.channel.ChannelMessageEvent
import com.github.twitch4j.chat.events.channel.SubscriptionEvent
import com.github.twitch4j.eventsub.events.ChannelFollowEvent
import com.github.twitch4j.eventsub.socket.IEventSubSocket
import com.github.twitch4j.eventsub.subscriptions.SubscriptionTypes
import com.github.twitch4j.pubsub.events.RewardRedeemedEvent
import com.theendercore.hydra.HydraMod.MODID
import com.theendercore.hydra.HydraMod.twitchClient
import com.theendercore.hydra.config.ModConfig
import com.theendercore.hydra.util.Methods.addChatMsg
import com.theendercore.hydra.util.darkGrayText
import com.theendercore.hydra.util.grayText
import com.theendercore.hydra.util.redText

object TwitchBot {

    @JvmField
    val ENABLED = false

    @JvmField
    var client: TwitchClient = TwitchClientBuilder.builder().build()

    @JvmField
    var credential: OAuth2Credential? = null

    @JvmField
    var eventSocket: IEventSubSocket? = null

    fun enable(): Int {
        return if (ENABLED) 0
        else {

            val config: ModConfig? = ModConfig.config
            credential = OAuth2Credential("twitch", config!!.oauthKey)

            if (config.username == "" || config.oauthKey == "") {
                addChatMsg(redText("command.$MODID.error.config"))
                return 0
            }
            if (twitchClient == null) {
                addChatMsg(darkGrayText("command.$MODID.connecting", config.username))
                twitchClient = TwitchClientBuilder.builder()
                    .withEnableHelix(true)
                    .withEnablePubSub(true)
                    .withEnableChat(true)
                    .withChatAccount(credential)
                    // New Api (yes its a pain)
                    .withEnableEventSocket(true)
                    .build()

                addChatMsg(darkGrayText("command.$MODID.connected"))
            } else {
                addChatMsg(darkGrayText("command.$MODID.connected.already", config.username))
                return 0
            }
            if (credential!!.userName == null) {
                addChatMsg(redText("command.$MODID.error.token.incorrect"))
                disable()
                return 0
            }
            if (config.extras) {
                val fetchedUsers = twitchClient!!.helix
                    .getUsers(credential!!.accessToken, null, listOf(config.username)).execute()

                config.broadcasterId = fetchedUsers.users[0].id
                config.save()

                twitchClient!!.pubSub.listenForChannelPointsRedemptionEvents(credential, config.broadcasterId)
                twitchClient!!.eventManager.onEvent(
                    RewardRedeemedEvent::class.java, EventListeners::rewardRedeemedListener
                )

                twitchClient!!.pubSub.listenForSubscriptionEvents(credential, config.broadcasterId)
                twitchClient!!.eventManager.onEvent(
                    SubscriptionEvent::class.java, EventListeners::subscriptionEventListener
                )

                eventSocket = twitchClient!!.getEventSocket()

                // create subscription
                eventSocket?.register(
                    SubscriptionTypes.CHANNEL_FOLLOW_V2
                        .prepareSubscription({ it.broadcasterUserId(config.broadcasterId).build() }, null)
                )

                // register event handler
                eventSocket?.eventManager?.onEvent(
                    ChannelFollowEvent::class.java, EventListeners::followingEventListener
                )

//                twitchClient!!.pubSub.listenForFollowingEvents(credential, config.broadcasterId)
//                twitchClient!!.eventManager.onEvent(FollowingEvent::class.java, EventListeners::followingEventListener)


                addChatMsg(grayText("command.$MODID.extras.enable"))
            }

            twitchClient!!.eventManager.onEvent(ChannelMessageEvent::class.java, EventListeners::channelMessageListener)

            1
        }

    }

    fun disable(): Int {
        return if (ENABLED) {
            client.pubSub.close()
            client.chat.close()
            client.close()
            addChatMsg(grayText("command.$MODID.disconnected"))
            1
        } else {
            addChatMsg(darkGrayText("command.$MODID.not_connected"))
            0
        }
    }
}